### PR TITLE
ToggleLockBranch: Add slack notification

### LINKED ISSUE #37: ToggleLockBranch: Add slack notifications
Send slack notifications when branch is locked or unlocked.

Similar feature is implemented in Python in https://github.com/SonarSource/gh-action_release/blob/6b2cd4455b7ec4a14b2a8ad728d20870174d1ea7/.github/workflows/main.yaml#L22

Steps needed:
* Pass `slack-token` as an optional parameter
* Pass `slack-channel` optional parameter
* Validate that token is present when the channel is set
* Add Slack SDK dependency
* Send message

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #37

Tested here: 
https://github.com/SonarSource/sonar-dotnet/pull/7518
And test runs from that branch:
https://github.com/SonarSource/sonar-dotnet/actions/runs/5412641167 - lock
https://github.com/SonarSource/sonar-dotnet/actions/runs/5412692988 - unlock

Results are in `#team-lang-dotnet-test-pavel` slack channel

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,12 +18,16 @@ All actions should be consumed from these branches.
 
 [AssignCardToSender](AssignCardToSender) - Assign Kanban card to sender of the event, mainly when moving card from `To do` column.
 
+[CopyLabelToPR](CopyLabelToPR) - Copy configured labels from all mentioned issues to PR.
```
Comment (pavel-mikula-sonarsource) [inline README.md:21]: left over from previous PR

File: ToggleLockBranch/ToggleLockBranch.ts [hunk 2]
```diff
@@ -15,9 +15,9 @@ class LockBranch extends GraphQLAction {
       const lockBranch = !rule.lockBranch;
       rule = await this.UpdateRule(rule.id, lockBranch);
       if (rule.lockBranch === lockBranch) {
-        this.log(
-          `Done: '${pattern}' was ${lockBranch ? 'locked' : 'unlocked and open for changes'}.`,
-        );
+        const message = `${this.repo.repo}: The branch \`${pattern}\` was ${lockBranch ? 'locked :ice_cube:' : 'unlocked and is now open for changes :sunny:'}`
```
Comment (pavel-mikula-sonarsource) [inline ToggleLockBranch/ToggleLockBranch.ts:18]: The `:ice_cube:` will not render in the GH action log. I'm fine with that to keep it simple

File: lib/OctokitAction.ts [hunk 3]
```diff
@@ -4,8 +4,9 @@ import { GitHub } from '@actions/github/lib/utils';
 import { Action } from './Action';
 import { RestEndpointMethods } from '@octokit/plugin-rest-endpoint-methods/dist-types/generated/method-types';
 import { IssueOrPR } from './IssueOrPR';
-import { Issue, ProjectCard, PullRequest } from './OctokitTypes';
+import { Issue, PullRequest } from './OctokitTypes';
 import { ProjectContent } from './ProjectContent';
+import fetch from 'node-fetch';
```
Comment (pavel-mikula-sonarsource) [inline lib/OctokitAction.ts:9]: I don't like how this looks, but I haven't found a better way

File: lib/OctokitAction.ts [hunk 4]
```diff
@@ -162,7 +163,47 @@ export abstract class OctokitAction extends Action {
...
+    } else {
+      this.log("Skip sending slack message, channel was not set.")
+    }
+  }
+
   private escapeRegex(string): string {
     return string.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&');
   }
+
+  private async sendSlackPost(url: string, jsonRequest: any): Promise<any> {
```
Comment (pavel-mikula-sonarsource) [inline lib/OctokitAction.ts:180]: I could have used `slack/web-api` npm package, but it has 3MB and 17 dependencies - that doesn't make sense to use.

File: ToggleLockBranch/ReadMe.md [hunk 5]
```diff
@@ -22,6 +28,12 @@ This parameter is optional.
 
 Default value: `master`
 
+### `slack-channel`
+
+Slack channel name or ID to send the notification to.
```
Comment (ilia-kebets-sonarsource) [inline ToggleLockBranch/ReadMe.md:33]:
```suggestion
Slack channel name or ID to send the notification to. When set, requires `slack-token` to be set as well.
```

File: ToggleLockBranch/action.yml [hunk 6]
```diff
@@ -4,10 +4,19 @@ inputs:
...
+  slack-token:
+    description: 'Slack auth Token with chat-write and chat-write.public scope. This is needed only when slack-channel parameter is also specified.'
+    required: false
+    default: ''
   branch-pattern:
     description: 'Pattern of the branch protection rule to modify.'
     required: false
     default: 'master'
+  slack-channel:
+    description: 'Slack channel name or ID to send the notification to.'
```
Comment (ilia-kebets-sonarsource) [inline ToggleLockBranch/action.yml:16]:
```suggestion
    description: 'Slack channel name or ID to send the notification to. When set, requires `slack-token` to be set as well.'
```

File: lib/OctokitAction.ts (hunk 3 above)
Comment (ilia-kebets-sonarsource) [reply lib/OctokitAction.ts:9]: you're following the examples from the doc, so it looks fine to me: https://www.npmjs.com/package/node-fetch

File: lib/OctokitAction.ts [hunk 7]
```diff
@@ -162,7 +163,47 @@ export abstract class OctokitAction extends Action {
     return result;
   }
 
+  protected async sendSlackMessage(text: string): Promise<void> {
+    const channel = this.getInput("slack-channel");
+    if (channel) {
+      this.log("Sending Slack message");
```
Comment (ilia-kebets-sonarsource) [inline lib/OctokitAction.ts:169]: you should run prettier, since it enforces single quotes

**Review by ilia-kebets-sonarsource — APPROVED** [review]
see comments about stylistic changes, otherwise LGTM!

File: lib/OctokitAction.ts (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply lib/OctokitAction.ts:169]: > 'prettier' is not recognized as an internal or external command,
> operable program or batch file.

Can you make it run automagically as part of the build?

**Pull request merged**